task-context.md

```diff
@@ -0,0 +1,191 @@
+when you return `NotReady`, or else you might never be `poll`ed again. The
+trick to the API is that there is an implicit `Task` argument to `Future::poll`
+which is passed using thread-local storage. This `Task` argument must be
+`notify`d by the `Future` in order for the task to awaken and `poll` again.
+It's essential to remember to do this, and to ensure that the right schedulings
+occur on every code path.
+
+[explicit task issue]: https://github.com/alexcrichton/futures-rs/issues/129
+
+## Reasoning about code
+
+The implicit `Task` argument makes it difficult to tell which functions
+aside from `poll` will use the `Task` handle to schedule a wakeup. That has a
+few implications.
+
+First, it's easy to accidentally call a function that will attempt to access
+the task outside of a task context. Doing so will result in a panic, but it 
+would be better to detect this mistake statically.
+
+Second, and relatedly, it is hard to audit a piece of code for which calls
+might involve scheduling a wakeup--something that is critical to get right
+in order to avoid "lost wakeups", which are far harder to debug than an
+explicit panic.
+
+## Debuggability
+
+Today's implicit TLS task system provides us with fewer "hooks" for adding debugging
+tools to the task system. Deadlocks and lost wakeups are some of the trickiest things
+to track down with futures code today, and it's possible that by using a "two stage"
+system like the one proposed in this RFC, we will have more room for adding debugging 
+hooks to track the precise points at which queing for wakeup occurred, and so on.
+
+More generally, the explicit `Context` argument and distinct `Waker` type provide greater
+scope for extensibility of the task system.
+
+## no_std
+
+While it's possible to use futures in `no_std` environments today, the approach
+is somewhat of a hack. Threading an explicit `&mut Context` argument makes no_std
+support entirely straightforward, which may also have implications for the futures
+library ultimately landing in libcore.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The `Future`, `Stream`, `AsyncRead`, and `AsyncWrite` traits have been modified
+to take explicit `task::Context` arguments. `task::Context` is a type which
+provides access to
+[task-locals](https://docs.rs/futures/*/futures/macro.task_local.html) and a `Waker`.
+
+In terms of the old `futures` task model, the `Waker` is analogous to the `Task`
+returned by [`task::current`][current], and the `wake` method is analogous to
+the old [`Task::notify`][notify].
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `task::Context` interface looks like this:
+
+```rust
+mod task {
+
+  /// The context for the currently running task.
+  pub struct Context {
+
+      /// Returns a `Waker` which can be used to wake up the current task.
+      pub fn waker(&self) -> Waker { ... }
```

Comment:
This is not Rust syntax, is `Context` a trait or a struct?

Reply:
Ah, sorry-- `Context` is a struct, this should be `pub struct Context { ... }`, `impl Context { ...`

Reply:
Possible alternative design:

```
pub fn waker(&self) -> (Waker, DidGetWaker)

#[must_use = "You must eventually call wake to allow the future to be polled again"]
pub struct Waker {
}

pub enum Async<T> {
    /// Represents that a value is immediately ready.
    Ready(T),

    /// Represents that a value is not ready yet, but may be so later.
    NotReady(DidGetWaker),
}
```

That way you:

- Must get the `Waker` in order to construct a new `NotReady`.
- Will get a warning if you never do anything with the `Waker`.